packages/swing-store/src/archiver.js

```diff
@@ -0,0 +1,44 @@
+import { finished as streamFinishedCallback, Readable } from 'node:stream';
+import { promisify } from 'node:util';
+import { createGzip } from 'node:zlib';
+import { withDeferredCleanup } from '@agoric/internal/src/node/utils.js';
+
+const streamFinished = promisify(streamFinishedCallback);
+
+/*
+ * @param {string} dirPath
+ * @param {object} powers
+ * @param {Pick<import('fs'), 'createWriteStream' | 'mkdirSync' | 'renameSync'>} powers.fs
+ * @param {Pick<import('path'), 'join'>} powers.path
+ * @param {Pick<import('tmp'), 'fileSync'>} powers.tmp
+ */
+export const makeArchiveTranscript = (dirPath, powers) => {
+  const { fs, path, tmp } = powers;
+  fs.mkdirSync(dirPath, { recursive: true });
+  const archiveTranscript = (spanName, entries) => {
+    const destPath = path.join(dirPath, `${spanName}.gz`);
+    return withDeferredCleanup(async addCleanup => {
+      const {
+        name: tmpName,
+        fd,
+        removeCallback,
+      } = tmp.fileSync({
+        prefix: spanName,
+        postfix: '.gz',
+        detachDescriptor: true,
+      });
+      addCleanup(() => removeCallback());
+      const writer = fs.createWriteStream('', { fd, flush: true });
+      const gzip = createGzip();
+      gzip.pipe(writer);
+      const reader = Readable.from(entries);
```

Comment:
I am feeling conflicted about the async shape of this, and the impact it has on the rest of the swing-store. Brian seemed interested in keeping things sync.

Previous discussion on this topic at https://github.com/Agoric/agoric-sdk/pull/8693#discussion_r1442125026, captured as #8739. With the caveat that here we get to write out as a stream as well, so even the output doesn't need to be kept as a whole in memory.

Reply:
I agree, but this function _really_ wants to be async, for both the deferred cleanups and for following the existing pattern established by export. The best alternative I can think of is having swing-store synchronously aggregate data into a Buffer and send that over, but on balance that seemed worse.